pylhc/lsa_to_madx.py

```diff
@@ -0,0 +1,241 @@
+Lines starting with a ``#`` character will be ignored.
+
+For instance the following **knobs.txt** file gathers a 2017 LHC configuration:
+
+.. code-block:: none
+
+    # Nonlinear Corrections
+    LHCBEAM/2017_IRNL_IR1a4	0.5
+    LHCBEAM/2017_IRNL_IR1b3_couplFD	1.0
+    LHCBEAM/2017_IRNL_a3b3_tuneFD 1.0
+    LHCBEAM/2017_IRNL_b4 1.0
+    LHCBEAM/2017_NL_IR1_30cm_kcssxr1 1.0
+    LHCBEAM/2018_IRNL_IR1a3 1.0
+    LHCBEAM/2018_IRNL_IR1b3 1.0
+    LHCBEAM/2018_IRNL_IR5a3 1.0
+    LHCBEAM/IR5_a4_2018 1.0
+    # Local correctors close to IP
+    LHCBEAM/2017_ATS_Inj_LocalCoupling 1.0
+    LHCBEAM/2017_ATS_LocalCorrection 1.0
+    # Chromatic Coupling
+    LHCBEAM/chromatic_coupling_b1_40cm_ctpps2_2017_v2 -1.0
+    LHCBEAM/chromatic_coupling_b2_40cm_ctpps2_2017 -1.0
+    # Beta-beat
+    LHCBEAM/b1_global_beta_beating_40cm_ctpps_2017_v2 1.0
+    LHCBEAM/b2_global_beta_beating_40cm_ctpps_2017 1.0
+    LHCBEAM/2017_beam1_beta_CrossingAngleCompensation 1.0
+    LHCBEAM/2017_30cm_GlobalCorr_wXing_Beam2_v2 1.0
+    LHCBEAM/B2_2017_global_correction_Xing 1.0
+    # Total Phase advance
+    LHCBEAM/2017_B2_MQT_total_phase_correction 1.0
+    LHCBEAM/B1_2017_MQT_total_phase_correction 1.0
+    # Arc-by-arc coupling
+    LHCBEAM/global_coup_knob_all_correctors	-1.0
+
+One can provide this file with the corresponding optics and get a ``MAD-X`` file for each defined knob with the appropriate trim value.
+In this case many files will be created.
+The call would be:
+
+.. code-block:: bash
+
+    python -m pylhc.lsa_to_madx \\
+        --optics R2017aT_A30C30A10mL300_CTPPS2 \\
+        --file knobs.txt
+
+Hint: the knobs active at a given time can be retrieved with the `~pylhc.machine_settings_info` script. 
+"""
+import argparse
+
+from pathlib import Path
+from typing import Dict
+
+import tfs
+
+from omc3.utils import logging_tools
+from omc3.utils.contexts import timeit
+
+from pylhc.data_extract.lsa import LSAClient
+
+LOG = logging_tools.get_logger(__name__)
+
+# ----- Helper functions ----- #
+
+
+def parse_knobs_and_trim_values_from_file(knobs_file: Path) -> Dict[str, float]:
+    """
+    Parses a file for LSA knobs and their trim values. Each line should be a knob name
+    following by a number of the trim value. If no value is written, it will be defaulted
+    to ``1.0``. Lines starting with a ``#`` are ignored.
+
+    Args:
+        knobs_file (~pathlib.Path): Path to the file with definitions.
+
+    Returns:
+        A `dict` with as keys the parsed knob names and as values their associated trims.
+    """
+    knob_lines = [line for line in Path(knobs_file).read_text().splitlines() if not line.startswith("#")]
+    results = {}
+
+    for line in knob_lines:
+        knob = line.split()[0]  # this is the knob name
+        try:  # catch the trim value, which is a number after the knob name
+            trim = float(line.split()[1])
+            results[knob] = trim
+        except IndexError:  # the is no number, we default the trim value to 1.0
```

Comment:
the trim value is not a number?

Reply:
It is, but the script accepts no value provided and will default to 1